### DISCUSSION TITLE
Importing an async function to server-side component from file that contains `useReducer` import

### DISCUSSION THREAD

**Discussion by justAnArthur** [vercel/next.js#47362] [Help] [answered]
### Summary

Hey)

I have built package via `ViteJs`. I want to import and use in `NextJs`'s app folder. I am importing an async function in server-side component to pass to client-side component and use in context. 

But when I have that import - `NextJs` gives me an error, coz that file I have imported from contains client-side logic:
```
./..\hooks\dist\reservation.js
ReactServerComponentsError:

You're importing a component that needs useReducer. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.

   ,-[C:\_\work\camasys-web-link\packages\hooks\dist\reservation.js:1:1]
 1 | import It, { createContext as pS, useRef as Hf, useState as Bf, createElement as z1, useContext as al, useEffect as ys, useLayoutEffect as Fj, useReducer as NT, useCallback as vS, useDebugValue as Tk, useMemo as ob } from "react";
   :                                                                                                                                                ^^^^^^^^^^
 2 | import { c as ru, E as np, g as on } from "./index-98f11477.js";
 3 | var Kh = {}, Lj = {
 4 |   get exports() {
   `----

Maybe one of these should be marked as a client entry with "use client":
  ./..\hooks\dist\reservation.js
  ./app\layout.js
  ```

What I actually do:
```
import {getStaticProps} from "camasys-web-provider/reservation";
....
```

I thought that because of having built by es-modules `NextJs` will take only that function.

Could you advise a solution of any kind ? Thank you very much 🙃.

**Comment by icyJoseph** [comment, marked as answer]
- https://beta.nextjs.org/docs/rendering/server-and-client-components#third-party-packages

I think because the way bundling works, and the way client directive applies (at module level), you cannot really import a function that shared module with a client hook, such as useReducer.

Is this async function exclusive to the server? I guess not, since it is bundled with useReducer. You could re-export it with something like:

```js
// create a new file just to re-export 😅 
'use client'

import {getStaticProps} from "camasys-web-provider/reservation";

export { getStaticProps };
```

Or any variation on that.

**Reply by justAnArthur** [reply]
That function must be called on the server, because uses authorization into.

But this trick with re-export might work as well. I'll try it, thanks.

Nextjs with those client and server components make this kind of hard to use external libraries.

**Reply by icyJoseph** [reply]
I think it won't work then... you might have to move that function into its own file within your library, so that you can pick it from the Server Component, without touching a client side module.
